draft-wang-ppm-dap-taskprov.md

```diff
@@ -56,76 +60,333 @@ collector). Thus it is up to each deployment of DAP to securely implement its
+        case time-interval: Empty;
+        case fixed-size: uint32 max_batch_size;
+    }
+} QueryConfig;
 ~~~
 
+The purpose of `TaskConfig` is to define all parameters that are necessary
+for configuring an aggregator. It includes all the fields to be
+associated with a task (see task configuration in
+{{?DAP=I-D.draft-ietf-ppm-dap-01}}.). Besides, `TaskConfig` also includes fields
+that useful for configuring a task in-band:
+
+* An opaque `task_info` that is specific to a task. For e.g. this can be a
+string describing the purpose of this task. It can also be the DAP task ID
+chosen by task author.
+
+* An opaque `vdaf_config` that contains any VDAF specific parameters for the
+chosen `vdaf_type`. The aggregators MUST pass `vdaf_config` to VDAF initialiser,
+based on the chosen `vdaf_type`.
+
+The codepoints for standardized (V)DAFs are listed below:
+
+~~~
+/* Codepoint for each standardized VDAF. Defined in
+ I-D.draft-irtf-cfrg-vdaf-02 */
+enum {
+    Prio3Aes128Count(0x00000000),
+    Prio3Aes128Sum(0x00000001),
+    Prio3Aes128Histogram(0x00000002),
+    Poplar1Aes128(0x00001000),
+    (2^32-1)
+} VdafType;
+~~~
+
+`VdafConfig` is not specified in this document, instead it should be defined by
+each VDAF implementation, for example, the simplest VDAF config for Prio3 can
+be defined as:
+
+~~~
+struct {
+    select (vdaf_type) {
+        case Prio3Aes128Count: Empty;
+        case Prio3Aes128Sum: uint8 bits;
+        case Prio3Aes128Histogram: uint32 buckets;
+    }
+} VdafConfig;
+~~~
+
+> OPEN ISSUE: Should DP parameters be defined as a different "dimension" to
+> VDAF, given that it's likely various DP mechanisms can be applied to any VDAF.
+
+The definition of Time, Duration, Url, QueryType follow those in
+{{?DAP=I-D.draft-ietf-ppm-dap-01}}.
+
+Note that `TaskConfig` does not encode all of the parameters required for the
+aggregator to run a task. In particular, parameters that should not be known to
+clients, like the VDAF verification key, the collector HPKE configuration, and
+whatever assets are required for HTTP request authentication are still
+established out-of-band.
 
 # Client Behavior
 
 > TODO(wangshan) Say how the Client constructs the extension body. Does the
 > extension change anything else about the upload flow?
 
+The client should know whether `task-prov` extension will be used and all
+parameters required for `TaskConfig` prior to constructing the extension
+body, either out-of-band from aggregators, or from information already saved on
+client.
+
+To use this extension, the client should first decide what is the task ID, see
+{{construct-task-id}}.
+
+Client constructs this extension during the upload flow, after hpke config
+(see update flow and hpke-config in {{?DAP=I-D.draft-ietf-ppm-dap-01}}.).
+Client typically sets `extension_type` to `task-prov` codepoint in
+`ReportMetadata`'s extension field, and save the encoded `TaskConfig` in
+`extension_data` field.
+
+## Construct task ID {#construct-task-id}
+
```

Comment:
> Task ID can be constructed
from `task_info` and any other part of task configuration, as long as the
generated task ID is determinstic and stay consistent across all parties. For
example, the task ID can be a SHA256 hash of the entire serialized
`TaskConfig`. 

The spec needs to be declaritive here. Instead of saying "here's one way one might compute the task ID", spell it out explicitly. For example:

> When the "task-prov" extension is used, the task ID is computed as follows:
> 
> ~~~
> task_id = SHA-256(task_config)
> ~~~

where `task_config` is the payload of the extension sent by the client.